File: NET/Libraries/Treehopper.Libraries/RegisterManager.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Treehopper.Libraries
{
    public class RegisterManager
    {
        protected List<Register> _registers = new List<Register>();
        protected SMBusDevice _dev;
        protected bool multiRegisterAccess;

        public RegisterManager(SMBusDevice dev, bool multiRegisterAccess)
        {
            _dev = dev;
            this.multiRegisterAccess = multiRegisterAccess;
        }

        public async Task Read(Register register)
        {
            var data = await _dev.ReadBufferData((byte) register.Address, register.Width).ConfigureAwait(false);
            register.SetBytes(data);
        }

        public async Task ReadRange(Register start, Register end)
        {
            if (multiRegisterAccess)
            {
                var count = (end.Address + end.Width) - start.Address;
                var bytes = await _dev.ReadBufferData((byte)start.Address, count).ConfigureAwait(false);
                int i = 0;
                foreach (var reg in _registers.Where(reg => reg.Address >= start.Address && reg.Address <= end.Address))
                {
                    reg.SetBytes(bytes.Skip(i).Take(reg.Width).ToArray());
                    i += reg.Width;
                }
            }
            else
            {
                foreach (var reg in _registers.Where(reg => reg.Address >= start.Address && reg.Address <= end.Address))
                {
                    await Read(reg);
                }
            }

        }

        public Task Write(Register register)
        {
            return _dev.WriteBufferData((byte) register.Address, register.GetBytes());
        }

        public async Task WriteRange(Register start, Register end)
        {
            if (multiRegisterAccess)
            {
                List<byte> bytes = new List<byte>();
                foreach (var reg in _registers.Where(reg => reg.Address >= start.Address && reg.Address <= end.Address))
                {
                    bytes.AddRange(reg.GetBytes());
                }
                await _dev.WriteBufferData((byte) start.Address, bytes.ToArray()).ConfigureAwait(false);
            }
            else
            {
                foreach (var reg in _registers.Where(reg => reg.Address >= start.Address && reg.Address <= end.Address))
                {
                    await Write(reg).ConfigureAwait(false);
                }
            }



        }
    }
}

File: C++/API/src/libraries/RegisterManager.cpp
#include <algorithm>
#include "Libraries/RegisterManager.h"
#include "Libraries/Register.h"

namespace Treehopper {
    namespace Libraries {
        RegisterManager::RegisterManager(SMBusDevice &dev, bool multiRegisterAccess)
			: dev(dev), 
			multiRegisterAccess(multiRegisterAccess)
		{

        }

        void RegisterManager::write(Register &reg) {
            dev.writeBufferData((uint8_t) reg.address, reg.getBytes());
        }

        void RegisterManager::writeRange(Register &start, Register &end) {
            if (multiRegisterAccess) {
                auto count = (end.address + end.width) - start.address;
                std::vector<uint8_t> bytes;

                for (auto &reg : registers) {
                    if (reg->address < start.address)
                        continue;

                    if (reg->address > end.address)
                        break;

                    auto data = reg->getBytes();
                    bytes.insert(bytes.end(), data.begin(), data.end());
                }

                dev.writeBufferData((uint8_t) start.address, bytes);
            } else {
                for (auto &reg : registers) {
                    if (reg->address < start.address)
                        continue;

                    if (reg->address > end.address)
                        break;

                    write(*reg);
                }
            }
        }

        void RegisterManager::read(Register &reg) {
            auto data = dev.readBufferData((uint8_t) reg.address, (size_t) reg.width);
            reg.setBytes(data);
        }

        void RegisterManager::readRange(Register &start, Register &end) {
            if (multiRegisterAccess) {
                auto count = (end.address + end.width) - start.address;
                auto bytes = dev.readBufferData((uint8_t) start.address, (size_t) count);
                int i = 0;

                for (auto &reg : registers) {
                    if (reg->address < start.address)
                        continue;

                    if (reg->address > end.address)
                        break;

                    std::vector<uint8_t> slice(bytes.begin() + i, bytes.begin() + i + reg->width);
                    reg->setBytes(slice);
                    i += reg->width;
                }
            } else {
                for (auto &reg : registers) {
                    if (reg->address < start.address)
                        continue;

                    if (reg->address > end.address)
                        break;

                    read(*reg);
                }
            }
        }
    }
}